File: dyd/roots/libraries/dhall-codegen-snapshot-tests/dyd/assets~fixture=person+target=py-codec-json/expected.py
from __future__ import annotations

from collections.abc import Awaitable, Callable
from datetime import date, datetime, time
from typing import Annotated, Any, Literal, Protocol, TypeAlias

from pydantic import BaseModel, TypeAdapter

# used to simulate intersection type
def _check_all_of(*adapters: TypeAdapter[Any]):
    def check(v: Any) -> Any:
        for ta in adapters:
            ta.validate_python(v)  # raises on failure
        return v
    return check


class PersonContact(BaseModel):# contact details
    phone_number : str
    email : (str) | None

class Person(BaseModel):
    appointment_time : time
    binary_data : str
    birth_date : str
    contact : PersonContact
    created_at : str
    date_of_birth : date
    destination_ip : str
    friends : list[Person]
    id : str
    name : str
    retention_period : str
    source_ip : str
    token : str

import base64
import math
import re
from datetime import date, datetime, time
from ipaddress import IPv4Address, IPv6Address, ip_address
from typing import Any
from urllib.parse import urlparse
from uuid import UUID


class CodecError(ValueError):
    def __init__(self, operation: str, path: str, message: str):
        super().__init__(f"{operation} error at {path}: {message}")
        self.operation = operation
        self.path = path
        self.message = message


def _fail(operation: str, path: str, message: str) -> None:
    raise CodecError(operation, path, message)


def _field(path: str, name: str) -> str:
    return f"{path}.{name}"


def _index(path: str, index: int) -> str:
    return f"{path}[{index}]"


def _object(operation: str, value: Any, path: str) -> dict[str, Any]:
    if not isinstance(value, dict) or any(not isinstance(key, str) for key in value):
        _fail(operation, path, "expected object")
    return value


def _array(operation: str, value: Any, path: str) -> list[Any]:
    if not isinstance(value, list):
        _fail(operation, path, "expected array")
    return value


def _string(operation: str, value: Any, path: str) -> str:
    if not isinstance(value, str):
        _fail(operation, path, "expected string")
    return value


def _number(operation: str, value: Any, path: str, kind: str) -> int | float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        _fail(operation, path, "expected number")
    if not math.isfinite(value):
        _fail(operation, path, "expected finite number")
    if kind in ("natural", "integer") and not isinstance(value, int):
        _fail(operation, path, "expected integer")
    if kind == "natural" and value < 0:
        _fail(operation, path, "expected natural number")
    return value


def _text(operation: str, value: Any, path: str, variant: str) -> str:
    value = _string(operation, value, path)
    if variant == "none":
        return value
    if variant == "email" and re.fullmatch(r"[^@\s]+@[^@\s]+\.[^@\s]+", value):
        return value
    if variant == "url":
        parsed = urlparse(value)
        if parsed.scheme and parsed.netloc:
            return value
    if variant == "base64":
        try:
            base64.b64decode(value, validate=True)
            return value
        except ValueError:
            pass
    if variant == "base64url":
        try:
            if re.fullmatch(r"[A-Za-z0-9_-]*={0,2}", value):
                base64.b64decode(value, altchars=b"-_", validate=True)
                return value
        except ValueError:
            pass
    if variant == "e164" and re.fullmatch(r"\+[1-9][0-9]{1,14}", value):
        return value
    if variant == "ipv4":
        try:
            if isinstance(ip_address(value), IPv4Address):
                return value
        except ValueError:
            pass
    if variant == "ipv6":
        try:
            if isinstance(ip_address(value), IPv6Address):
                return value
        except ValueError:
            pass
    if variant == "isoDate":
        try:
            date.fromisoformat(value)
            return value
        except ValueError:
            pass
    if variant == "isoDateTime":
        try:
            datetime.fromisoformat(value.replace("Z", "+00:00"))
            return value
        except ValueError:
            pass
    if variant == "isoTime":
        try:
            time.fromisoformat(value)
            return value
        except ValueError:
            pass
    if variant == "isoDuration" and re.fullmatch(r"P(?=\d|T\d)(?:\d+Y)?(?:\d+M)?(?:\d+W)?(?:\d+D)?(?:T(?:\d+H)?(?:\d+M)?(?:\d+(?:\.\d+)?S)?)?", value):
        return value
    if variant == "uuid":
        try:
            UUID(value)
            return value
        except ValueError:
            pass
    if variant.startswith("literal:") and value == variant[8:]:
        return value
    _fail(operation, path, f"invalid {variant} text")


def _any(operation: str, value: Any, path: str) -> Any:
    if value is None or isinstance(value, (bool, str)):
        return value
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return _number(operation, value, path, "none")
    if isinstance(value, list):
        for index, item in enumerate(value):
            _any(operation, item, _index(path, index))
        return value
    if isinstance(value, dict) and all(isinstance(key, str) for key in value):
        for key, item in value.items():
            _any(operation, item, _field(path, key))
        return value
    _fail(operation, path, "expected JSON value")


def _time(operation: str, value: Any, path: str, variant: str) -> Any:
    if variant == "none":
        if operation == "encode":
            if not isinstance(value, datetime): _fail(operation, path, "expected datetime")
            return value.isoformat()
        try:
            return datetime.fromisoformat(_string(operation, value, path).replace("Z", "+00:00"))
        except ValueError:
            _fail(operation, path, "expected ISO date-time")
    if variant == "date":
        if operation == "encode":
            if not isinstance(value, date) or isinstance(value, datetime): _fail(operation, path, "expected date")
            return value.isoformat()
        try:
            return date.fromisoformat(_string(operation, value, path))
        except ValueError:
            _fail(operation, path, "expected ISO date")
    if variant == "time":
        if operation == "encode":
            if not isinstance(value, time): _fail(operation, path, "expected time")
            return value.isoformat()
        try:
            return time.fromisoformat(_string(operation, value, path))
        except ValueError:
            _fail(operation, path, "expected ISO time")
    return _text(operation, value, path, "isoDuration")



def _encode_PersonContact_at(value: PersonContact, path: str) -> Any:
    result = {}
    result["phone_number"] = None
    result["phone_number"] = _text("encode", value.phone_number, _field(path, "phone_number"), "e164")
    if value.email is not None:
        result["email"] = _text("encode", value.email, _field(path, "email"), "email")
    return result

def _decode_PersonContact_at(input: Any, path: str) -> PersonContact:
    object = _object("decode", input, path)
    result = {}
    if "phone_number" not in object:
        _fail("decode", _field(path, "phone_number"), "missing required field")
    result["phone_number"] = _text("decode", object["phone_number"], _field(path, "phone_number"), "e164")
    if "email" in object:
        result["email"] = _text("decode", object["email"], _field(path, "email"), "email")
    else:
        result["email"] = None
    return PersonContact.model_construct(**result)

def encode_PersonContact(value: PersonContact) -> Any:
    return _encode_PersonContact_at(value, "$")

def decode_PersonContact(input: Any) -> PersonContact:
    return _decode_PersonContact_at(input, "$")


def _encode_Person_at(value: Person, path: str) -> Any:
    result = {}
    result["appointment_time"] = None
    result["appointment_time"] = _time("encode", value.appointment_time, _field(path, "appointment_time"), "time")
    result["binary_data"] = None
    result["binary_data"] = _text("encode", value.binary_data, _field(path, "binary_data"), "base64")
    result["birth_date"] = None
    result["birth_date"] = _text("encode", value.birth_date, _field(path, "birth_date"), "isoDate")
    result["contact"] = None
    result["contact"] = _encode_PersonContact_at(value.contact, _field(path, "contact"))
    result["created_at"] = None
    result["created_at"] = _text("encode", value.created_at, _field(path, "created_at"), "isoDateTime")
    result["date_of_birth"] = None
    result["date_of_birth"] = _time("encode", value.date_of_birth, _field(path, "date_of_birth"), "date")
    result["destination_ip"] = None
    result["destination_ip"] = _text("encode", value.destination_ip, _field(path, "destination_ip"), "ipv6")
    result["friends"] = None
    if not isinstance(value.friends, list):
        _fail("encode", _field(path, "friends"), "expected list")
    result["friends"] = []
    for index, entry in enumerate(value.friends):
        converted_Personx = _encode_Person_at(entry, _index(_field(path, "friends"), index))
        result["friends"].append(converted_Personx)
    result["id"] = None
    result["id"] = _text("encode", value.id, _field(path, "id"), "uuid")
    result["name"] = None
    result["name"] = _text("encode", value.name, _field(path, "name"), "none")
    result["retention_period"] = None
    result["retention_period"] = _time("encode", value.retention_period, _field(path, "retention_period"), "duration")
    result["source_ip"] = None
    result["source_ip"] = _text("encode", value.source_ip, _field(path, "source_ip"), "ipv4")
    result["token"] = None
    result["token"] = _text("encode", value.token, _field(path, "token"), "base64url")
    return result

def _decode_Person_at(input: Any, path: str) -> Person:
    object = _object("decode", input, path)
    result = {}
    if "appointment_time" not in object:
        _fail("decode", _field(path, "appointment_time"), "missing required field")
    result["appointment_time"] = _time("decode", object["appointment_time"], _field(path, "appointment_time"), "time")
    if "binary_data" not in object:
        _fail("decode", _field(path, "binary_data"), "missing required field")
    result["binary_data"] = _text("decode", object["binary_data"], _field(path, "binary_data"), "base64")
    if "birth_date" not in object:
        _fail("decode", _field(path, "birth_date"), "missing required field")
    result["birth_date"] = _text("decode", object["birth_date"], _field(path, "birth_date"), "isoDate")
    if "contact" not in object:
        _fail("decode", _field(path, "contact"), "missing required field")
    result["contact"] = _decode_PersonContact_at(object["contact"], _field(path, "contact"))
    if "created_at" not in object:
        _fail("decode", _field(path, "created_at"), "missing required field")
    result["created_at"] = _text("decode", object["created_at"], _field(path, "created_at"), "isoDateTime")
    if "date_of_birth" not in object:
        _fail("decode", _field(path, "date_of_birth"), "missing required field")
    result["date_of_birth"] = _time("decode", object["date_of_birth"], _field(path, "date_of_birth"), "date")
    if "destination_ip" not in object:
        _fail("decode", _field(path, "destination_ip"), "missing required field")
    result["destination_ip"] = _text("decode", object["destination_ip"], _field(path, "destination_ip"), "ipv6")
    if "friends" not in object:
        _fail("decode", _field(path, "friends"), "missing required field")
    result["friends"] = []
    for index, entry in enumerate(_array("decode", object["friends"], _field(path, "friends"))):
        converted_Personx = _decode_Person_at(entry, _index(_field(path, "friends"), index))
        result["friends"].append(converted_Personx)
    if "id" not in object:
        _fail("decode", _field(path, "id"), "missing required field")
    result["id"] = _text("decode", object["id"], _field(path, "id"), "uuid")
    if "name" not in object:
        _fail("decode", _field(path, "name"), "missing required field")
    result["name"] = _text("decode", object["name"], _field(path, "name"), "none")
    if "retention_period" not in object:
        _fail("decode", _field(path, "retention_period"), "missing required field")
    result["retention_period"] = _time("decode", object["retention_period"], _field(path, "retention_period"), "duration")
    if "source_ip" not in object:
        _fail("decode", _field(path, "source_ip"), "missing required field")
    result["source_ip"] = _text("decode", object["source_ip"], _field(path, "source_ip"), "ipv4")
    if "token" not in object:
        _fail("decode", _field(path, "token"), "missing required field")
    result["token"] = _text("decode", object["token"], _field(path, "token"), "base64url")
    return Person.model_construct(**result)

def encode_Person(value: Person) -> Any:
    return _encode_Person_at(value, "$")

def decode_Person(input: Any) -> Person:
    return _decode_Person_at(input, "$")

File: dyd/roots/libraries/dhall-codegen-snapshot-tests/dyd/assets~fixture=person+target=ts-codec-json/expected.ts
export type PersonContact =
	/** contact details */
	{
		phone_number :
			/** E.164 telephone number */
			string;
		email ?:
			/** contact email */
			string;
	};

export type Person =
	{
		appointment_time :
			/** ISO 8601 local time */
			string;
		binary_data :
			/** RFC 4648 Base64 */
			string;
		birth_date :
			/** ISO 8601 calendar date */
			string;
		contact :
			/** contact details */
			PersonContact;
		created_at :
			/** RFC 3339 date-time */
			string;
		date_of_birth :
			/** date of birth */
			Date;
		destination_ip :
			/** IPv6 address */
			string;
		friends :
			/** friends */
			Array<
				Person
			>;
		id :
			/** RFC 4122 UUID */
			string;
		name :
			/** full name */
			string;
		retention_period :
			/** ISO 8601 duration */
			string;
		source_ip :
			/** IPv4 address */
			string;
		token :
			/** RFC 4648 Base64url */
			string;
	};

export type JsonValue =
  | null
  | boolean
  | number
  | string
  | JsonValue[]
  | { [key: string]: JsonValue };

export class CodecError extends Error {
  constructor(
    public readonly operation: "encode" | "decode",
    public readonly path: string,
    message: string,
  ) {
    super(operation + " error at " + path + ": " + message);
    this.name = "CodecError";
  }
}

export class UnsupportedCodecError extends CodecError {
  constructor(operation: "encode" | "decode", path: string, schema: string) {
    super(operation, path, "unsupported schema " + schema);
    this.name = "UnsupportedCodecError";
  }
}

const hasOwn = (value: object, key: string): boolean =>
  Object.prototype.hasOwnProperty.call(value, key);

const fail = (operation: "encode" | "decode", path: string, message: string): never => {
  throw new CodecError(operation, path, message);
};

const unsupported = (operation: "encode" | "decode", path: string, schema: string): never => {
  throw new UnsupportedCodecError(operation, path, schema);
};

const pathField = (path: string, field: string): string => path + "." + field;
const pathIndex = (path: string, index: number): string => path + "[" + index + "]";

const asObject = (operation: "encode" | "decode", value: unknown, path: string): Record<string, unknown> => {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    return fail(operation, path, "expected object");
  }

  return value as Record<string, unknown>;
};

const asArray = (operation: "encode" | "decode", value: unknown, path: string): unknown[] => {
  if (!Array.isArray(value)) {
    return fail(operation, path, "expected array");
  }

  return value;
};

const asString = (operation: "encode" | "decode", value: unknown, path: string): string => {
  if (typeof value !== "string") {
    return fail(operation, path, "expected string");
  }

  return value;
};

const asNumber = (operation: "encode" | "decode", value: unknown, path: string): number => {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    return fail(operation, path, "expected finite number");
  }

  return value;
};

const validateText = (variant: string, value: string, operation: "encode" | "decode", path: string): string => {
  const matches = (pattern: RegExp): boolean => pattern.test(value);

  if (variant.startsWith("literal:")) {
    const literal = variant.slice("literal:".length);
    return value === literal ? value : fail(operation, path, "expected literal " + JSON.stringify(literal));
  }

  switch (variant) {
    case "none": return value;
    case "email": return matches(/^[^@\s]+@[^@\s]+\.[^@\s]+$/) ? value : fail(operation, path, "expected email");
    case "url":
      try {
        new URL(value);
        return value;
      } catch {
        return fail(operation, path, "expected URL");
      }
    case "base64": return matches(/^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/) ? value : fail(operation, path, "expected Base64");
    case "base64url": return matches(/^[A-Za-z0-9_-]*={0,2}$/) ? value : fail(operation, path, "expected Base64url");
    case "e164": return matches(/^\+[1-9][0-9]{1,14}$/) ? value : fail(operation, path, "expected E.164 phone number");
    case "ipv4": return matches(/^(?:25[0-5]|2[0-4][0-9]|1?[0-9]{1,2})(?:\.(?:25[0-5]|2[0-4][0-9]|1?[0-9]{1,2})){3}$/) ? value : fail(operation, path, "expected IPv4 address");
    case "ipv6": return value.includes(":") ? value : fail(operation, path, "expected IPv6 address");
    case "isoDate": return matches(/^\d{4}-\d{2}-\d{2}$/) ? value : fail(operation, path, "expected ISO date");
    case "isoDateTime": return matches(/^\d{4}-\d{2}-\d{2}T/) ? value : fail(operation, path, "expected RFC 3339 date-time");
    case "isoDuration": return matches(/^P(?=\d|T\d)/) ? value : fail(operation, path, "expected ISO duration");
    case "isoTime": return matches(/^\d{2}:\d{2}:\d{2}(?:\.\d+)?$/) ? value : fail(operation, path, "expected local ISO time");
    case "uuid": return matches(/^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i) ? value : fail(operation, path, "expected UUID");
    default: return fail(operation, path, "unknown text variant " + variant);
  }
};

const asJsonObject = (operation: "encode" | "decode", value: unknown, path: string): Record<string, unknown> => {
  const object = asObject(operation, value, path);
  const prototype = Object.getPrototypeOf(object);

  if (prototype !== Object.prototype && prototype !== null) {
    return fail(operation, path, "expected plain object");
  }

  return object;
};

const validateAny = (operation: "encode" | "decode", value: unknown, path: string): JsonValue => {
  if (value === null || typeof value === "boolean" || typeof value === "string") return value as null | boolean | string;
  if (typeof value === "number") return asNumber(operation, value, path);
  if (Array.isArray(value)) {
    for (let index = 0; index < value.length; index += 1) {
      validateAny(operation, value[index], pathIndex(path, index));
    }
    return value as JsonValue[];
  }

  const object = asJsonObject(operation, value, path);
  for (const key in object) {
    if (hasOwn(object, key)) validateAny(operation, object[key], pathField(path, key));
  }
  return object as { [key: string]: JsonValue };
};

const encodeAny = (value: unknown, path: string): JsonValue => validateAny("encode", value, path);

const decodeAny = (value: unknown, path: string): JsonValue => validateAny("decode", value, path);

const encodeBoolean = (value: unknown, path: string): JsonValue =>
  typeof value === "boolean" ? value : fail("encode", path, "expected boolean");

const decodeBoolean = (value: unknown, path: string): boolean =>
  typeof value === "boolean" ? value : fail("decode", path, "expected boolean");

const encodeNumber = (variant: string, value: unknown, path: string): JsonValue => {
  const number = asNumber("encode", value, path);
  if (variant === "natural" && (!Number.isInteger(number) || number < 0)) fail("encode", path, "expected natural number");
  if (variant === "integer" && !Number.isInteger(number)) fail("encode", path, "expected integer");
  return number;
};

const decodeNumber = (variant: string, value: unknown, path: string): number => {
  const number = asNumber("decode", value, path);
  if (variant === "natural" && (!Number.isInteger(number) || number < 0)) fail("decode", path, "expected natural number");
  if (variant === "integer" && !Number.isInteger(number)) fail("decode", path, "expected integer");
  return number;
};

const encodeText = (variant: string, value: unknown, path: string): JsonValue =>
  validateText(variant, asString("encode", value, path), "encode", path);

const decodeText = (variant: string, value: unknown, path: string): string =>
  validateText(variant, asString("decode", value, path), "decode", path);

const encodeTime = (variant: string, value: unknown, path: string): JsonValue => {
  if (variant === "none" || variant === "date") {
    if (!(value instanceof Date) || Number.isNaN(value.getTime())) return fail("encode", path, "expected valid Date");
    return variant === "date" ? value.toISOString().slice(0, 10) : value.toISOString();
  }

  return validateText(variant === "time" ? "isoTime" : "isoDuration", asString("encode", value, path), "encode", path);
};

const decodeTime = (variant: string, value: unknown, path: string): unknown => {
  const text = asString("decode", value, path);
  const textVariant = variant === "none" ? "isoDateTime" : variant === "date" ? "isoDate" : variant === "time" ? "isoTime" : "isoDuration";
  validateText(textVariant, text, "decode", path);

  if (variant === "none") {
    const result = new Date(text);
    return Number.isNaN(result.getTime()) ? fail("decode", path, "invalid date-time") : result;
  }

  if (variant === "date") return new Date(text + "T00:00:00.000Z");
  return text;
};



function encodePersonContactAt(value: PersonContact, path: string): JsonValue {
	const object = asObject("encode", value, path);
	const result: { [key: string]: JsonValue } = {};
	result["phone_number"] = encodeText("e164", object["phone_number"], pathField(path, "phone_number"));
	if (hasOwn(object, "email") && object["email"] !== undefined) result["email"] = encodeText("email", object["email"], pathField(path, "email"));
	return result;
}

function decodePersonContactAt(input: unknown, path: string): PersonContact {
	const object = asObject("decode", input, path);
	const result: { [key: string]: unknown } = {};
	result["phone_number"] = decodeText("e164", object["phone_number"], pathField(path, "phone_number"));
	if (hasOwn(object, "email") && object["email"] !== null) result["email"] = decodeText("email", object["email"], pathField(path, "email"));
	return result as PersonContact;
}

export function encodePersonContact(value: PersonContact): JsonValue {
	return encodePersonContactAt(value, "$");
}

export function decodePersonContact(input: unknown): PersonContact {
	return decodePersonContactAt(input, "$");
}

function encodePersonAt(value: Person, path: string): JsonValue {
	const object = asObject("encode", value, path);
	const result: { [key: string]: JsonValue } = {};
	result["appointment_time"] = encodeTime("time", object["appointment_time"], pathField(path, "appointment_time"));
	result["binary_data"] = encodeText("base64", object["binary_data"], pathField(path, "binary_data"));
	result["birth_date"] = encodeText("isoDate", object["birth_date"], pathField(path, "birth_date"));
	result["contact"] = encodePersonContactAt(object["contact"] as PersonContact, pathField(path, "contact"));
	result["created_at"] = encodeText("isoDateTime", object["created_at"], pathField(path, "created_at"));
	result["date_of_birth"] = encodeTime("date", object["date_of_birth"], pathField(path, "date_of_birth"));
	result["destination_ip"] = encodeText("ipv6", object["destination_ip"], pathField(path, "destination_ip"));
	result["friends"] = ((input: unknown, path: string): JsonValue[] => {
		const entries = asArray("encode", input, path);
		const result: JsonValue[] = new Array(entries.length);
		for (let index = 0; index < entries.length; index += 1) {
			result[index] = encodePersonAt(entries[index] as Person, pathIndex(path, index));
		}
		return result;
	})(object["friends"], pathField(path, "friends"));
	result["id"] = encodeText("uuid", object["id"], pathField(path, "id"));
	result["name"] = encodeText("none", object["name"], pathField(path, "name"));
	result["retention_period"] = encodeTime("duration", object["retention_period"], pathField(path, "retention_period"));
	result["source_ip"] = encodeText("ipv4", object["source_ip"], pathField(path, "source_ip"));
	result["token"] = encodeText("base64url", object["token"], pathField(path, "token"));
	return result;
}

function decodePersonAt(input: unknown, path: string): Person {
	const object = asObject("decode", input, path);
	const result: { [key: string]: unknown } = {};
	result["appointment_time"] = decodeTime("time", object["appointment_time"], pathField(path, "appointment_time"));
	result["binary_data"] = decodeText("base64", object["binary_data"], pathField(path, "binary_data"));
	result["birth_date"] = decodeText("isoDate", object["birth_date"], pathField(path, "birth_date"));
	result["contact"] = decodePersonContactAt(object["contact"], pathField(path, "contact"));
	result["created_at"] = decodeText("isoDateTime", object["created_at"], pathField(path, "created_at"));
	result["date_of_birth"] = decodeTime("date", object["date_of_birth"], pathField(path, "date_of_birth"));
	result["destination_ip"] = decodeText("ipv6", object["destination_ip"], pathField(path, "destination_ip"));
	result["friends"] = ((input: unknown, path: string): unknown[] => {
		const entries = asArray("decode", input, path);
		const result: unknown[] = new Array(entries.length);
		for (let index = 0; index < entries.length; index += 1) {
			result[index] = decodePersonAt(entries[index], pathIndex(path, index));
		}
		return result;
	})(object["friends"], pathField(path, "friends"));
	result["id"] = decodeText("uuid", object["id"], pathField(path, "id"));
	result["name"] = decodeText("none", object["name"], pathField(path, "name"));
	result["retention_period"] = decodeTime("duration", object["retention_period"], pathField(path, "retention_period"));
	result["source_ip"] = decodeText("ipv4", object["source_ip"], pathField(path, "source_ip"));
	result["token"] = decodeText("base64url", object["token"], pathField(path, "token"));
	return result as Person;
}

export function encodePerson(value: Person): JsonValue {
	return encodePersonAt(value, "$");
}

export function decodePerson(input: unknown): Person {
	return decodePersonAt(input, "$");
}